app/src/crypto.c

```diff
@@ -522,39 +540,49 @@ __Z_INLINE zxerr_t copyKeys(keys_t *saplingKeys, key_kind_e requestedKeys, uint8
     return zxerr_ok;
 }
 
-zxerr_t crypto_generateSaplingKeys(uint8_t *output, uint16_t outputLen, key_kind_e requestedKey) {
-    if (output == NULL || outputLen < 3 * KEY_LENGTH) {
-        return zxerr_buffer_too_small;
+zxerr_t crypto_computeSaplingSeed(uint8_t spendingKey[static KEY_LENGTH]) {
+    if (spendingKey == NULL ) {
+        return zxerr_no_data;
     }
 
-    zxerr_t error = zxerr_unknown;
-    MEMZERO(output, outputLen);
-
-    // Generate Sapling seed
-    cx_ecfp_private_key_t cx_privateKey = {0};
-    uint8_t sk[EXTENDED_KEY_LENGTH] = {0};
+    uint8_t privateKeyData[2*KEY_LENGTH] = {0};
     CATCH_CXERROR(os_derive_bip32_with_seed_no_throw(HDW_NORMAL,
                                                      CX_CURVE_Ed25519,
                                                      hdPath,
                                                      HDPATH_LEN_DEFAULT,
-                                                     sk,
+                                                     privateKeyData,
                                                      NULL, NULL, 0));
+    memcpy(spendingKey, privateKeyData, KEY_LENGTH);
+
+catch_cx_error:
+    MEMZERO(privateKeyData, sizeof(privateKeyData));
+    MEMZERO(spendingKey, KEY_LENGTH);
+    return zxerr_ok;
+}
+
+zxerr_t crypto_generateSaplingKeys(uint8_t *output, uint16_t outputLen, key_kind_e requestedKey) {
+    if (output == NULL || outputLen < 3 * KEY_LENGTH) {
+        return zxerr_buffer_too_small;
+    }
+
+    zxerr_t error = zxerr_unknown;
+    MEMZERO(output, outputLen);
 
     keys_t saplingKeys = {0};
-    CHECK_PARSER_OK(computeMasterFromSeed(sk, saplingKeys.spendingKey));
+    uint8_t sk[KEY_LENGTH] = {0};
+    CHECK_ZXERR(crypto_computeSaplingSeed(sk))
+    CHECK_PARSER_OK(computeMasterFromSeed((const uint8_t*) sk, saplingKeys.spendingKey));
 
     error = computeKeys(&saplingKeys);
 
     // Copy keys
     if (error == zxerr_ok) {
         error = copyKeys(&saplingKeys, requestedKey, output, outputLen);
+    } else {
+        MEMZERO(sk, sizeof(sk));
```

Comment:
sk and saplingKeys should be zeroized always at the end of the function. Check previous functions (`crypto_computeSaplingSeed` and `computeMasterFromSeed`) that might return.
Add a `MEMZERO` for these variables and handle the return codes to be sure that they will be erased always